lib/settings/settings.js

```diff
@@ -0,0 +1,85 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict'
+
+const assert  = require('assert')
+
+const KEY = Symbol()
+const POLL_INTERVAL = Symbol()
+
+module.exports = (config, mc, log) => {
+
+
+  // A sentinel error for signaling that the result was invalid/missing,
+  // and we should try to push our own representation to memcached.
+  class PushOnMissing extends Error {}
+
+  // An abstract class that stores options in memcached.
+  //
+  // Others extend this class to hotload options from memcached.
+  class Settings {
+
+    constructor(key) {
+      assert(typeof key === 'string')
+      this[KEY] = key
+    }
+
+    // subclasses should provide their own implementation
+    setAll(value) {
+      return this
+    }
+
+    // subclasses should provide their own implementation
+    validate(value) {
+      return value
+    }
+
+    // Pushes `this` as JSON to `key`
+    //
+    // Customize what is pushed with a toJSON method.
+    push() {
+      log.info({ op: this[KEY] + '.push' })
+      return mc.setAsync(this[KEY], this, 0)
+        .then(() => this.refresh())
+    }
+
+    refresh(options) {
+      log.info({ op: this[KEY] + '.refresh' })
+      let result = mc.getAsync(this[KEY]).then(value => this.validate(value))
+
+      if (options && options.pushOnMissing) {
+        result = result.catch(err => {
+          if (err instanceof PushOnMissing) {
```

Comment:
IIUC, the "if missing" part of this kind of happens implicitly:

* We try to the value from memcache, but it's missing, so we get `null` or `undefined` or something
* We call `this.validate(value)` with the missing result
* The implementation of `this.validate` does some type-checking and errors out with `PushOnMissing` because it's not an arrow/object/what-have-you

Is it worth adding an explicit check for "the item was missing" here in the base class, in case a future subclass doesn't know that it have to perform this check inside `validate()`?